instrumentation/opentelemetry_ecto/lib/opentelemetry_ecto.ex

```diff
@@ -55,58 +55,54 @@ defmodule OpentelemetryEcto do
     total_time = measurements.total_time
     end_time = :opentelemetry.timestamp()
     start_time = end_time - total_time
-    database = repo.config()[:database]
 
-    url =
-      case repo.config()[:url] do
-        nil ->
-          # TODO: add port
-          URI.to_string(%URI{scheme: "ecto", host: repo.config()[:hostname]})
+    %{
+      hostname: hostname,
+      username: username,
+      database: database,
+      port: port
+    } =
+      [hostname: nil, username: nil, database: nil, port: nil]
+      |> Keyword.merge(repo.config())
+      |> Keyword.merge(Ecto.Repo.Supervisor.parse_url(repo.config()[:url] || ""))
+      |> Keyword.take([:hostname, :username, :database, :port])
+      |> Enum.into(%{})
 
-        url ->
-          url
-      end
+    url =
+      "ecto://#{username}@#{hostname}:#{port}/#{database}"
+      |> URI.parse()
+      |> URI.to_string()
 
     span_name =
       case Keyword.fetch(config, :span_prefix) do
         {:ok, prefix} -> prefix
         :error -> Enum.join(event, ".")
-      end <> if source != nil, do: ":#{source}", else: ""
+      end
+      |> maybe_append_source(source)
 
     time_unit = Keyword.get(config, :time_unit, :microsecond)
 
-    db_type =
-      case type do
-        :ecto_sql_query -> :sql
-        _ -> type
-      end
-
-    # TODO: need connection information to complete the required attributes
-    # net.peer.name or net.peer.ip and net.peer.port
-    base_attributes = %{
-      "db.type": db_type,
-      "db.statement": query,
-      source: source,
-      "db.instance": database,
-      "db.url": url,
-      "total_time_#{time_unit}s": System.convert_time_unit(total_time, :native, time_unit)
-    }
+    base_attributes =
+      %{
+        "db.type": if(type == :ecto_sql_query, do: :sql, else: type),
+        "db.statement": query,
+        "db.instance": database,
+        "db.url": url,
+        "total_time_#{time_unit}s": System.convert_time_unit(total_time, :native, time_unit)
+      }
+      |> maybe_add(:source, source)
```

Comment:
Why is this considered optional at this point? We want this attribute to always be present for consistency even if it is an empty string.

Reply:
Currently It will either be `nil` or some string value. I was converting `nil` into an "optional" attribute. Do we want to always include it? Should we convert `nil` to `""`?